### PR TITLE
8219014: (bf) Add absolute bulk put methods which accept a source Buffer

### PR DESCRIPTION
**Pull request by bplb**
Please review this proposed fix. The previous incarnation of this review was most recently in mail threads [1, 2] under the old Hg SCM. The implementation changes proposed here are derived from the final (.02) patch posted in [1]. The corresponding CSR is [3].

The essence of the change is to move the bulk put buffer core functionality to a new method putBuffer(), modify the existing relative bulk put buffer method to call it, and add a new absolute bulk put buffer method which also calls it. The existing test for bulk put buffer functionality is modified to verify the new API.

This change builds on the work completed for JDK-5029431 (bulk put and get using an array) and JDK-8245121 (relative bulk put buffer with overlapping regions).

[1] http://mail.openjdk.java.net/pipermail/nio-dev/2020-June/007340.html
[2] http://mail.openjdk.java.net/pipermail/nio-dev/2020-July/007344.html
[3] https://bugs.openjdk.java.net/browse/JDK-8248431

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8219014](https://bugs.openjdk.java.net/browse/JDK-8219014): (bf) Add absolute bulk put methods which accept a source Buffer

### Reviewers
 * [Paul Sandoz](https://openjdk.java.net/census#psandoz) (@PaulSandoz - **Reviewer**) ⚠️ Review applies to 1cfc8aba6c7e351a2ba64ae737de050001a53fdb
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to 1cfc8aba6c7e351a2ba64ae737de050001a53fdb

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/969/head:pull/969`
`$ git checkout pull/969`

### CODE REVIEW

**Comment by mjpt777** [comment]
The other method implementations which do a absolute put or get do not modify the position. This makes them more useful in a multi-threaded context as different threads can be updating different ranges of the buffer concurrently. To replace Unsafe we need absolute index methods that do not update position.

**Comment by cl4es** [comment]
> The other method implementations which do a absolute put or get do not modify the position. This makes them more useful in a multi-threaded context as different threads can be updating different ranges of the buffer concurrently. To replace Unsafe we need absolute index methods that do not update position.

Thanks for pointing out the importance of not depending on `position` for these methods to be as helpful as they can be. 

I checked and the new methods here (`put(int, $Type$Buffer, int, int)`) does not modify or read `position`. I think the diff here makes it look like `position` is in use for the new methods since the diff gets bungled up with the preceding, position-relative `put($Type$Buffer)` method, and that this played some tricks.

**Review by PaulSandoz — APPROVED** [review]
I like the way this patch builds on prior work and reuse, including the test. It gives us confidence, since this is a tricky area to get right, esp. with edge cases like `StringCharBuffer`.

**Comment by PaulSandoz** [comment]
@mjpt777 adding to Alan's comment, we are looking for feedback on the memory API, especially in real-world performance sensitive use-cases (see design [doc](https://github.com/openjdk/panama-foreign/blob/foreign-jextract/doc/panama_memaccess.md), latest proposed incubating [api](http://cr.openjdk.java.net/~mcimadamore/8254162_v1/javadoc/jdk/incubator/foreign/package-summary.html)).

File: src/java.base/share/classes/java/nio/X-Buffer.java.template [hunk 1]
```diff
@@ -956,6 +956,76 @@ public abstract class $Type$Buffer
...
+     * {@code limit() - index} and {@code src.limit() - offset}
+     *
+     * @return This buffer
+     *
+     * @throws IndexOutOfBoundsException
+     * If the preconditions on the {@code index}, {@code offset}, and
+     * {@code length} parameters do not hold
+     *
+     * @throws ReadOnlyBufferException
+     * If this buffer is read-only
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/nio/X-Buffer.java.template:1013]: Can you align the parameter and throws descriptions so they are consistent with the other methods?

File: test/jdk/java/nio/Buffer/BulkPutBuffer.java [hunk 2]
```diff
@@ -294,6 +323,41 @@ boolean equals(Buffer src, Buffer dst) throws Throwable {
         return args.toArray(Object[][]::new);
     }
 
+    private static void expectThrows(Class exClass,
+        Assert.ThrowingRunnable r) {
```
Comment (AlanBateman) [inline test/jdk/java/nio/Buffer/BulkPutBuffer.java:327]: Spurious line break in method declaration, also exClass should be Class<?>

**Review by AlanBateman — APPROVED** [review]

**Comment by bplb** [comment]
Thanks for the comments. A second reviewer on the CSR would not hurt.

**Pull request closed without merging**